### PR TITLE
fix: stand down on the legacy image key instead of writing the v2 home beside it

### PR DESCRIPTION
**Pull request by KrasimirKralev**
TASK-743 — the image migration in `scripts/gateway-pre-start.sh` writes the
image model under the v2 home while the legacy home is still on the box. Plus
the cloud-voice downgrade half reading only one home, and the two small items
the #754 and #755/#756 reviewers asked to be folded in.

Read the "what this PR does NOT do" paragraph before the rest: the card files
this as link 1 of the customer incident, and it is not — measured against the
core, the block's write never caused that outage and the blocked box is not
rescued here.

## What the customer sees

**The owner's own image model, thrown away by us.** On a v2 core the two homes
for the image-generation pick are `agents.defaults.imageGenerationModel`
(legacy) and `agents.defaults.mediaModels.image`. This block read the legacy key
as a fallback but treated the shapes that resolve no model by ITS rule — `{}`,
`{"primary": ""}`, `{"fallbacks": []}`, a JSON `null`, and a bare string, which
the core resolves as a model even though this block does not — as an empty slot,
and claimed the v2 home.

That matters because of what the core's own migration does with a home that is
already taken. `migrateFinalLayoutRenames` (read out of the pinned bundle):

```js
if (!Object.hasOwn(defaults, legacyKey)) continue;
if (mediaModels[canonicalKey] === void 0) mediaModels[canonicalKey] = defaults[legacyKey];
else changes.push(`Removed agents.defaults.${legacyKey} (…already set).`);
delete defaults[legacyKey];            // both branches
```

So writing the v2 home first does not strand the box — the legacy key is deleted
either way — it makes the core DISCARD the owner's value instead of carrying it
across. Standing down leaves the one shape the core can migrate, and makes one
fewer write into a config it is refusing anyway.

**Measured on the pinned core** (`openclaw 2026.8.1 (ea80657)`, on the OpenClaw
box, `openclaw config validate --json` in a throwaway
`HOME`/`OPENCLAW_CONFIG_PATH`/`OPENCLAW_STATE_DIR` under `/tmp`, `OPENCLAW_HOME`
unset, scratch removed afterwards — the real config was never opened):

| config | verdict |
|---|---|
| `imageGenerationModel: {"primary": ""}` **and** `mediaModels.image` — what beta writes | `rc=1`, `agents.defaults: Unrecognized key: "imageGenerationModel"` |
| `imageGenerationModel: {"primary": ""}` alone — what this fix leaves | `rc=1`, the same issue |
| `mediaModels.image` alone — after the core's migration | `rc=0`, `{"valid":true,…,"warnings":[]}` |

**What this PR does NOT do, said plainly.** The second row is the point twice
over. `agents.defaults` is `.strict()`, so the legacy key ALONE is already exit
78 — this block's write never caused the outage it was filed under. And the only
box that reaches the new guard is one whose `doctor --fix` did not run the
migration (the block at the top of this script runs it whenever the core refuses
the config; a non-empty legacy `exec-approvals.json` is what stops it). **That
box stays dark with this change and without it, so link 1 of TASK-737 is not
closed here.** Performing the move in this file instead — which it already does
for `agents.defaults.tools` and `systemPromptSuffix` 1,300 lines above — is a
real option and wants its own card; it is a bigger decision than a stand-down,
and #751 deliberately did not take it for the sibling key.

**A Pro box paying for a refused round trip on every spoken reply.** The
cloud-voice DOWNGRADE arm read only `cfg["tts"]` under v2. A box that was Max
under a 2026.7 core wrote its stamped entry into `messages.tts`, and a v2 core
that has not run its migration still has it there — so the arm found nothing,
left our own dead entry in place, and the Voice panel went on calling the cloud
voice configured while every utterance bought a 403.

## The fix

Both halves are #751's shape, and the second is #751's own arm reaching one home
further.

- **A stand-down keyed on the legacy key's PRESENCE**, not its contents, around
  the image SLOT write. `in`, not `.get() is not None`, because a key whose
  value is `null` is present and is refused too. Standing down is never
  permanent: the next boot after `doctor --fix` sees the key moved and writes
  the slot as always.
- **One predicate for "this slot holds a decision", used by #755's take-back
  arm.** That arm asked `is not None`, so a leftover EMPTY legacy key —
  exactly the shapes the guard above treats as holding nothing — vetoed the
  whole stand-down, and a box whose credential the proxy had permanently refused
  kept the image path declared (the TASK-727 storm, 6,554 refused calls in
  twelve hours). The new predicate is deliberately WIDER for the delete than for
  the upsert: it counts a bare string, because the core coerces one and this is
  the only place in the file that destroys configuration.
- **The sibling in-flight guard one screen up had the same `null` hole the new
  comment says `in` was chosen for.** `tools.media.audio.models: null` is
  present and refused, and `.get("models") is not None` read it as absent — so
  the audio block seeded `tools.media.models` beside the surviving legacy key.
  Fixed with `"models" in _audio` plus a seed guard, with its own case.
- **Narrower than #751's guard, deliberately.** There the whole block wrote ONE
  object into one of two homes, so standing down had to mean writing nothing.
  Here the block writes two independent things: the `models[]` ROW, which has
  one home on both cores and whose repair (a missing `name`) is what keeps a box
  bootable, and the SLOT, which is the half with two homes. Only the slot stands
  down, so the boot after the migration lands has nothing left to do.
- **The downgrade arm reads both homes on v2** and deletes our stamped entry
  from whichever one holds it. Deleting from the legacy home is safe where
  WRITING to it is not: a removal cannot create the dual-home shape the guard
  above stands down over, and it leaves the core's own migration less to move.
  The stamp is still the only thing that opens that door — it is the one place
  in the file that destroys configuration — so the extra home widens WHERE the
  arm looks, never WHAT it may delete.
- **#755's refusal/take-back arm is untouched.** It is in the same block and was
  read first; it removes each home on its own already (`_legacy_is_ours` /
  `_v2_is_ours`), which is the same rule pointing the other way, and the new
  guard sits on the upsert arm only.

## The sibling: the other writer of the same slot

`grep` for the slot found a second writer with the identical hole, and it is the
half an owner reaches by pressing **Save** rather than by rebooting:
`buildClawboxAiImageOps` in `src/app/setup-api/ai-models/configure/route.ts`
honours the legacy key as "already configured" only when it NAMES a model, then
pushes `agents.defaults.mediaModels.image`. Same guard applied there, same
reason, with the `it.each` of empty shapes in `configure-images.test.ts` split
into "the key is absent → claim the slot" and "the key is there → stand down".
That file's helper also stopped spelling *absent* as `{ imageGenerationModel:
undefined }`: openclaw.json is JSON, so the distinction that decides this is
missing versus `null`, and a property explicitly set to `undefined` is neither.

## Sibling call sites, each cleared

- **Every `_clawbox_v2` switch in the boot script**, which is where a
  dual-home write can hide: `gateway.controlUi.allowInsecureAuth` /
  `dangerouslyDisableDeviceAuth` DELETE the retired keys under v2 (the opposite
  shape); the codex switches decide model refs and profiles, not homes;
  `tools.media.audio.models` already has its own `_audio_models_move_in_flight`
  stand-down; `messages.tts` is #751. The image slot was the last one without a
  guard.
- **Readers of the image slot** — `harness/capabilities.ts:174,243`,
  `harness/clawai-images.ts:23`, `harness/transport.ts:81`,
  `openclaw-config.ts:1173-1179` — read it or describe it, and none writes it;
  `openclaw-config.ts` already types both homes.
- **`askCoreToValidateConfig`'s killed-child guard** (extra (a) below): the
  other three readers of that flag in `updater.ts` are cleared with a reason —
  `isRetryableRemoteFailure:182` and `describeRemoteFailure:275` decide whether a
  git failure is worth retrying, and a buffer that overflowed is not a timeout
  and must not be retried into another overflow; `describeDoctorFailure:1305`
  and `startRootStep:2888` only choose the WORDING of a failure that is a
  failure either way. `apps/install/route.ts:225` is the same: `killed` picks a
  message, never a verdict. This one is the only site where the flag gates a
  PARSE, and the parse decides whether the owner's plugin entries are switched
  off.
- **`@/lib/config-store` mocks** (extra (b)): the new gate found seven suites
  with the hole, all fixed in this PR.

## Extras folded in, each with its own test

**(a) `src/lib/updater.ts` — the killed-validator guard did not fire on a
maxBuffer overflow.** Node kills a child whose stdout passes `maxBuffer` and
rejects with `code: "ERR_CHILD_PROCESS_STDIO_MAXBUFFER"` while `killed` and
`signal` are both `undefined`, so the two-flag test let a TRUNCATED buffer reach
the parser — and `coreReportedMissingPlugins` takes `warnings[]` off the payload
whatever the exit was. LOW (2 MB against a payload of tens of bytes), and the
cut is not always unparseable: a slice between the first `{` and the last `}` can
still be valid JSON carrying half a verdict.

**(b) `openclaw-config-mock-completeness.test.ts` now guards
`@/lib/config-store` too.** `@/lib/clawai-credential-refusal` reads and writes
the persisted ClawBox AI credential refusal, and both of its calls sit inside a
catch that answers a DEFAULT — so a factory missing `get` does not fail a suite,
it silently makes every case take the "no refusal on record" branch. #755 fixed
that by hand in three files; #756 opened it again in a fourth the same day.

**The scope is the hazard, not the import graph.** Nearly every route reaches
that reader transitively, and the whole-graph rule the sibling gate uses reports
FIFTY files — a list nobody acts on. What makes a suite able to take the wrong
answer is that its own subject calls the reader, so the gate is one hop: a test
that imports a module which imports `@/lib/clawai-credential-refusal` itself.
That found **seven** offenders, all fixed here.

One of those seven turned out to be a test passing because of the hole:
`configure-subscription-surface.test.ts`'s case for a save that cannot touch the
subscription surface at all asserted `not.toHaveBeenCalled()` over the WHOLE of
`fs/promises.readFile`, and it held only because the missing export threw inside
the save so the request never reached `forgetProviderEnumerations()` — which
reads `catalog-cache/_enumerations.json` whenever `models.mode` moved, and on a
box always has. The pin is narrowed to the per-provider surface cache, which is
what that file is about.

## RED → GREEN

**RED**, against beta with the new tests in place — 12 failures across all four
areas:

```
 ❯ src/tests/unit/gateway-pre-start-clawai-images.test.ts (80 tests | 8 failed)
 ❯ src/tests/unit/gateway-pre-start-clawai-speech.test.ts (46 tests | 2 failed)
 ❯ src/tests/unit/openclaw-config-mock-completeness.test.ts (6 tests | 1 failed)
 ❯ src/tests/unit/updater.test.ts (93 tests | 1 failed)

 FAIL  … a legacy image key the core has not migrated yet > stands down rather than
       writing the v2 home beside an empty primary
 AssertionError: expected { primary: 'openai/gpt-image-1-mini' } to be undefined

 FAIL  … taking the cloud voice back from wherever the entry actually is >
       takes back a stamped entry stranded in the legacy home
 AssertionError: expected false to be true

 FAIL  … acts on nothing a validator that OVERFLOWED its buffer had written
 AssertionError: expected true to be false      (byteplus was switched off)

 FAIL  … every @/lib/config-store mock … carries its store functions
 AssertionError: expected [ …(7) ] to deeply equal []
```

The first of those is the defect in one line: on beta the script writes
`mediaModels.image` while `imageGenerationModel` is still in the file.

**GREEN.** All four suites pass, and so does the whole unit project. `tsc
--noEmit` clean, ESLint clean on the changed files, `bash -n` clean on
`scripts/gateway-pre-start.sh`, and all 23 python heredocs in that script
compile (`compile()` per heredoc — the blocks are not files, so `py_compile`
needs the extraction).

## Both editions

`scripts/gateway-pre-start.sh` is OpenClaw's, and **the Hermes edition never
enters it** — proved from the shipped code rather than asserted:

- It runs as `ExecStartPre` of `clawbox-gateway.service`
  (`config/clawbox-gateway.service:61`, and `install-x64.sh:1088` writes the same
  line). On the `hermes` edition `step_edition_gateway_state` stops, disables,
  REMOVES and masks that unit; it is dropped from `EXPECTED_ACTIVE_SERVICES` and
  listed in `FOREIGN_EDITION_UNITS`, so `step_validate_services` fails if it ever
  comes back — pinned by `install-foreign-edition-teardown.test.ts`.
- The one other invoker is `runCurrentGatewayPreStart()` in `updater.ts`, reached
  only from `ensureGatewayHealthy()`, whose two callers are the `gateway_verify`
  step (`applies: () => !gatewayIsAbsent()`, i.e. not on `hermes`) and the
  OpenClaw-only update flow, which `startOpenclawUpdate()` refuses outright when
  `openclawIsAbsent()`. `updater.test.ts` already pins that an update on a Hermes
  box does not run `gateway_verify`.

`dual` DOES run it, correctly — that SKU runs both harnesses and this is
OpenClaw's own config. The `updater.ts` and route changes are edition-agnostic
and shared.

## Review

`clawbox-review` (one review agent, findings at `code-review-task743.md`):
FIX-FIRST, 2 HIGH / 4 MEDIUM / 3 LOW / 2 INFO — and the two HIGHs were about
this PR's own record rather than its code. The reviewer read
`migrateFinalLayoutRenames` out of the pinned bundle and ran `doctor --fix`
against an isolated home, and both of the claims the first draft rested on were
wrong: the legacy key is deleted on BOTH branches (no stranding), and
`agents.defaults` being `.strict()` means the legacy key alone is already the
exit 78 (this block never caused it). Every comment that carried those claims —
in the script, in the route, and in both test files — is rewritten to what was
measured, and the "does not rescue the blocked box" paragraph above exists
because of that finding.

Applied in full: the shared take-back predicate (MEDIUM-3), the `null` hole in
the sibling audio guard (MEDIUM-2), the route's untested owner-pick guard, which
every legacy-seeded case now short-circuits past (MEDIUM-1, two new cases
seeding `mediaModels.image` directly), the missing version-gate note on the
route (MEDIUM-4), `Object.hasOwn(null, …)` throwing and taking the provider row
with it (LOW-1 — a real bug this PR introduced), the corrected cause behind the
narrowed `readFile` assertion (LOW-2 — re-measured: `setProviderEnabled` inside
a try/catch reaches `readProviderStatus` -> `_enumerations.json`, and with `get`
missing the whole re-enable step was silently skipped), and the three stale
comments (LOW-3). Both INFOs are recorded in the run queue as their own cards.

## Bug classes

- *False success* — the image block reported (and persisted) a slot write that
  left the box unable to load its config at all. That is the change.
- *False failure* — the downgrade arm reporting "nothing of ours here" over an
  entry that was ours, one home over.
- *Probe-once* — none introduced; every read here is per boot.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved AI model migrations to preserve existing image, audio, and voice configurations across OpenClaw versions.
  * Prevented duplicate or premature configuration entries, including configurations with empty or null legacy values.
  * Improved cleanup of refused or outdated provider and voice configurations while preserving customer settings.
  * Prevented truncated configuration validation output from being processed as valid.

* **Tests**
  * Added regression coverage for migration, cleanup, downgrade, and validation edge cases.
  * Improved test coverage for configuration reads and writes.

### CODE REVIEW

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 1]
```diff
@@ -170,3 +170,90 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+    for (const importer of importers.get(caller) ?? []) exposed.add(importer);
+  }
+
+  const offenders: string[] = [];
+  for (const file of exposed) {
+    if (!/\.test\.tsx?$/.test(file)) continue;
+    const factory = mockCall(read.get(file)!, STORE_MODULE);
+    if (!factory) continue;
+    if (factory.includes("importActual") || factory.includes("importOriginal")) continue;
+    const missing = STORE_EXPORTS.filter((name) => !new RegExp(`\\b${name}\\s*[:,)]`).test(factory));
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Applied — and it was the failure mode a gate must not have.

Your reading is right: `\b${name}\s*[:,)]` misses `get() {}` and `async set() {}`, so a factory that is COMPLETE, written with object methods, would have been reported as missing and failed the build over correct code. A gate whose false positives land on correct code is one people switch off.

Taken with two additions rather than as a one-character edit:
- the expression is now a named `storeExportNamed(name)` helper with the reason on it, so the offender scan and the self-test cannot drift apart — they were two copies of the same literal before;
- the self-test gained the method form as a case, and a `getAll`/`getKnown` case in the other direction. Those two prefixes are live in this repo's factories, and `\b` anchoring the start is the only thing that stops them reading as `get` — worth pinning next to the change that widened the class.

Pushed as `24cef3ab`. The file is green (6 passed), the gate still reports 0 offenders across the 16 suites it watches, `tsc --noEmit` and ESLint are clean.

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 2]
```diff
@@ -170,3 +170,111 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+ *
+ * `get: vi.fn()`, `get,` (shorthand), `get)` (last entry) — and `get()` /
+ * `async get()`, the object-METHOD form. The last one is why the opening paren
+ * is in the class: without it a perfectly complete factory written with methods
+ * would be reported as missing, and a gate whose failure mode is a false
+ * positive is one nobody trusts. `\b` anchors the start, so `getAll:` and
+ * `getKnown:` — both live in this repo's factories — never match `get`.
+ */
+function storeExportNamed(name: string): RegExp {
+  return new RegExp(`\\b${name}\\s*[:,)(]`);
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Both shapes were real, and taking them seriously turned up a third. The matcher is gone; the gate now reads the members the factory actually exports.

`factoryExports()` finds the object the factory returns (`() => ({ … })`, or a block body's `return {`) and walks it: strings and template literals skipped whole, bracket depth counted, and a name taken only where a member may begin — straight after that object's `{`, or after a `,` at the same depth. `async`, and the `get`/`set` accessor keywords, are treated as modifiers when another identifier follows, so `async get() {}` is the member `get` and `get DATA_DIR() {}` is the member `DATA_DIR`. A factory whose returned object it cannot find is now REPORTED rather than skipped — the sibling gate in this file learned that a mock the scanner cannot see is the one failure a gate must not have.

Your two cases are pinned, and so is what the first accurate run found: seven COMPLETE factories reported as missing an export, and one as unparseable, because every factory in this repo explains itself between its members. An apostrophe in a comment ("the owner's picks") opened a string that swallowed the rest of the object, and a bare word after a comma read as the next member's name. Comments are consumed before the quote check now, with both comment forms in the self-test.

Known limit, stated rather than hidden: a regex literal containing a quote inside a factory would still desync the walk. None exists, and `test-timeout-hygiene.test.ts` documents the same hazard for its own walker.

Pushed as `082980a8`. The file is green, the gate still reports 0 offenders across the suites it watches, and the whole unit project passes (755 files, 12,536 tests).

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 3]
```diff
@@ -170,3 +170,242 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+ * Returns null when the factory's returned object cannot be found, and the
+ * caller REPORTS such a factory rather than skipping it: a shape this cannot
+ * read is a shape nobody has checked.
+ */
+function factoryExports(factory: string): Set<string> | null {
+  // `() => ({ … })` is the form every factory in this repo uses; `return {`
+  // covers a block-bodied one. Anything else is unparseable here by design.
+  const arrow = factory.indexOf("=>");
+  if (arrow < 0) return null;
+  const parenObject = factory.indexOf("({", arrow);
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Right again, and it is the same failure direction as the last one: the gate passing over the hole it exists for.

`returnedObjectStart()` now takes only two shapes — the DIRECT arrow-expression body (`=>` then `(` then `{`, each with trivia skipped) and a block body's own top-level `return`, found by walking the body at its own depth so a `return` inside a nested function or object is not mistaken for it. A parenthesised `return ({ … })` is read too. Anything else is `null`, and a `null` is reported as an offender rather than skipped.

Your case is in the self-test verbatim:

    () => { const d = ({ get: vi.fn(), set: vi.fn() }); return { set: d.set }; }

answers `["set"]` now, where taking the first `({` after the arrow answered `["get","set"]` and would have cleared a factory that omits `get`.

Trivia skipping is shared between that lookup and the member walk, so a comment sitting between `=>` and the object — or between `return` and its object — cannot desync one while the other reads it.

Pushed as the head of this branch. The file is green (6 passed), the gate still reports 0 offenders, and the whole unit project passes (755 files, 12,536 tests); `tsc --noEmit` and ESLint clean.

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 4]
```diff
@@ -170,3 +170,313 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+      i += 1;
+      while (i < factory.length && factory[i] !== quote) {
+        if (factory[i] === "\\") i += 1;
+        i += 1;
+      }
+      continue;
+    }
+    if (ch === "{" || ch === "[" || ch === "(") { depth += 1; continue; }
+    if (ch === "}" || ch === "]" || ch === ")") { depth -= 1; if (depth === 0) return -1; continue; }
+    if (depth !== 1 || ch !== "r" || !/^return\b/.test(factory.slice(i))) continue;
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Taken the second way you offer: a branch-dependent factory is now REPORTED as unreadable rather than answered from one of its returns.

Your case is exact — `() => { if (p) { return { set }; } return { get, set }; }` has a runtime path that omits `get`, and reading the unconditional return would clear it. That is the same failure direction as the two before it, so the walker no longer looks past depth: a `return` found anywhere below the block body's own level ends the scan with `null`, and `null` lands the file in the offender list with `unparseable factory` beside it, where a human reads it.

Reported rather than parsed, deliberately. Following every reachable return means deciding which are reachable, and a gate whose scanner needs a control-flow analysis is a gate with its own bugs. Every factory this repo actually writes is `() => ({ … })`; the block-bodied branch exists so an unusual one is not silently skipped, and the honest answer for an unusual one is "nobody has checked this".

The case is in the self-test next to the local-object one. Pushed as `6ea7a2eb`: file green, gate still 0 offenders, `tsc --noEmit` and ESLint clean.

On the suite: the last two full `--project unit` runs on this machine each reported four failures, a DIFFERENT four each time (`telegram/status`, `catalog-empty-enumeration`, `clawkeep-memory`, `root-exec-manifest`) — all files this diff neither touches nor imports, all green in isolation, and the identical tree passed 755 files / 12,536 tests twice earlier. Load average was 16.7 on a 12-core box at the time; this is the dev-PC load-flake family already recorded in the run notes, and the `test` job here is the arbiter.

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 5]
```diff
@@ -170,3 +170,325 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+      i += 1;
+      while (i < factory.length && factory[i] !== quote) {
+        if (factory[i] === "\\") i += 1;
+        i += 1;
+      }
+      continue;
+    }
+    if (ch === "{" || ch === "[" || ch === "(") { depth += 1; continue; }
+    if (ch === "}" || ch === "]" || ch === ")") { depth -= 1; if (depth === 0) return -1; continue; }
+    if (ch !== "r" || !/^return\b/.test(factory.slice(i))) continue;
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Both correct, and together they settled the design: the block-bodied branch is deleted rather than fixed a fourth time.

`myreturn` did match — the test only asked that the slice START with `return` and never that a token boundary preceded it. And `if (p) return { … };` is unbraced, so the `return` sits at the body's own depth and the depth test says nothing about whether it is conditional.

What those two have in common with the local-object case and the braced-conditional case before them is the point: reading `return { … }` out of a `{ … }` body is not a lookup, it is control flow. Four ways to answer from a path that is not the factory's only path, four review rounds, and every failure in the direction that CLEARS a factory nobody checked. A gate that needs a control-flow analysis to be right has bugs of its own.

So `returnedObjectStart` now reads exactly one shape — the direct arrow-expression body `() => ({ … })`, which is what every mock factory in this repo is written as — and answers -1 for everything else. -1 means the file is REPORTED with `unparseable factory` beside it, so an unusual factory gets a human rather than a guess. It costs nothing while no factory uses the other shape, and the day one does, the gate says so instead of pretending.

All six shapes are pinned as unreadable in the self-test: a helper passed by name, a plain block body, a local object before the return, a braced conditional return, your unbraced one, and `const myreturn = 1`.

Pushed as `850ed9e9`: the file is green, the gate still reports 0 offenders across the suites it watches, `tsc --noEmit` and ESLint clean, and the whole unit project passes — 755 files, 12,536 tests, on a quiet machine.

File: src/tests/unit/openclaw-config-mock-completeness.test.ts [hunk 6]
```diff
@@ -170,3 +170,325 @@ describe(`every ${MOCKED_MODULE} mock that can reach an ${NARROWED_EXPORT} narro
...
+    // A CONDITIONAL return makes the factory unreadable, not merely different.
+    // `() => { if (partial) return { set }; return { get, set }; }` has a path
+    // that omits an export, and answering from the unconditional one would
+    // clear it — the gate passing over the hole it exists for, for the third
+    // time in this review. Reported instead, so a human reads the factory.
+    if (depth !== 1) return -1;
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/openclaw-config-mock-completeness.test.ts]: Both correct, and together they settled the design: the block-bodied branch is deleted rather than fixed a fourth time.

`myreturn` did match — the test only asked that the slice START with `return` and never that a token boundary preceded it. And `if (p) return { … };` is unbraced, so the `return` sits at the body's own depth and the depth test says nothing about whether it is conditional.

What those two have in common with the local-object case and the braced-conditional case before them is the point: reading `return { … }` out of a `{ … }` body is not a lookup, it is control flow. Four ways to answer from a path that is not the factory's only path, four review rounds, and every failure in the direction that CLEARS a factory nobody checked. A gate that needs a control-flow analysis to be right has bugs of its own.

So `returnedObjectStart` now reads exactly one shape — the direct arrow-expression body `() => ({ … })`, which is what every mock factory in this repo is written as — and answers -1 for everything else. -1 means the file is REPORTED with `unparseable factory` beside it, so an unusual factory gets a human rather than a guess. It costs nothing while no factory uses the other shape, and the day one does, the gate says so instead of pretending.

All six shapes are pinned as unreadable in the self-test: a helper passed by name, a plain block body, a local object before the return, a braced conditional return, your unbraced one, and `const myreturn = 1`.

Pushed as `850ed9e9`: the file is green, the gate still reports 0 offenders across the suites it watches, `tsc --noEmit` and ESLint clean, and the whole unit project passes — 755 files, 12,536 tests, on a quiet machine.

**Comment by KrasimirKralev** [comment]
Merged into beta (fbaae33d) after final review MERGE-OK on this head and green CI on this head, all review threads closed. Deploy of beta to both boxes follows; box proof lands on TASK-727 TASK-737 .
